adapter/prometheus/prometheus.go

```diff
@@ -32,40 +38,185 @@ type (
 		once sync.Once
 	}
 
-	prom struct{}
+	prom struct {
+		metrics map[string]prometheus.Collector
+	}
 )
 
 var (
 	name = "prometheus"
 	desc = "Publishes prometheus metrics"
 	conf = &config.Params{}
+
+	charReplacer = strings.NewReplacer("/", "_", ".", "_", " ", "_")
 )
 
 // Register records the builders exposed by this adapter.
 func Register(r adapter.Registrar) {
-	r.RegisterMetricsBuilder(newFactory())
+	r.RegisterMetricsBuilder(newFactory(newServer(defaultAddr)))
 }
 
-func newFactory() *factory {
-	return &factory{adapter.NewDefaultBuilder(name, desc, conf), newServer(defaultAddr), sync.Once{}}
+func newFactory(srv server) *factory {
+	return &factory{adapter.NewDefaultBuilder(name, desc, conf), srv, sync.Once{}}
 }
 
 func (f *factory) Close() error {
 	return f.srv.Close()
 }
 
+// NewMetricsAspect provides an implementation for adapter.MetricsBuilder.
+// It can panic if there is a failure to register a metric from the list of
+// metric definitions provided.
 func (f *factory) NewMetricsAspect(env adapter.Env, cfg adapter.AspectConfig, metrics []adapter.MetricDefinition) (adapter.MetricsAspect, error) {
 	var serverErr error
 	f.once.Do(func() { serverErr = f.srv.Start(env.Logger()) })
 	if serverErr != nil {
 		return nil, fmt.Errorf("could not start prometheus server: %v", serverErr)
 	}
 
-	return &prom{}, nil
+	metricsMap := make(map[string]prometheus.Collector, len(metrics))
+	for _, m := range metrics {
+		switch m.Kind {
+		case adapter.Gauge:
+			metricsMap[m.Name] = mustRegisterOrGet(newGaugeVec(m.Name, m.Description, m.Labels))
```

Comment:
We have the ability to return an error here, why not change `mustRegisterOrGet` to `registerOrGet` and return an error instead of panicking?